src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -204,1024 +204,1045 @@ internal void OnBeforeWrite(TokenCacheNotificationArgs args)
+                            "Access token is expired.  IsExtendedLifeTimeEnabled=TRUE and ExtendedExpiresOn is not exceeded.  Returning the found cache entry. " +
+                            GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+
+                        msalAccessTokenCacheItem.IsExtendedLifeTimeToken = true;
+                        return msalAccessTokenCacheItem;
+                    }
+
                     requestParams.RequestContext.Logger.Info(
-                        "Access token is expired.  IsExtendedLifeTimeEnabled=TRUE and ExtendedExpiresOn is not exceeded.  Returning the found cache entry. " +
+                        "Access token has expired or about to expire. " +
                         GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
-
-                    msalAccessTokenCacheItem.IsExtendedLifeTimeToken = true;
-                    return msalAccessTokenCacheItem;
                 }
 
-                requestParams.RequestContext.Logger.Info(
-                    "Access token has expired or about to expire. " +
-                    GetAccessTokenExpireLogMessageContent(msalAccessTokenCacheItem));
+                return null;
             }
-
-            return null;
         }
-    }
 
-    private string GetAccessTokenExpireLogMessageContent(MsalAccessTokenCacheItem msalAccessTokenCacheItem)
-    {
-        return string.Format(
-            CultureInfo.InvariantCulture,
-            "[Current time ({0}) - Expiration Time ({1}) - Extended Expiration Time ({2})]",
-            DateTime.UtcNow,
-            msalAccessTokenCacheItem.ExpiresOn,
-            msalAccessTokenCacheItem.ExtendedExpiresOn);
-    }
+        private ICollection<MsalAccessTokenCacheItem> FilterToAuthoritySpecifiedByAuthenticationRequest(
+            AuthenticationRequestParameters requestParams, ICollection<MsalAccessTokenCacheItem> tokenCacheItems)
+        {
+            if (tokenCacheItems.Count <= 1)
+            {
+                return tokenCacheItems;
+            }
 
-    internal async Task<MsalRefreshTokenCacheItem> FindRefreshTokenAsync(AuthenticationRequestParameters requestParams)
-    {
-        using (ServiceBundle.TelemetryManager.CreateTelemetryHelper(requestParams.RequestContext.TelemetryRequestId, requestParams.RequestContext.ClientId,
-            new CacheEvent(CacheEvent.TokenCacheLookup) { TokenType = CacheEvent.TokenTypes.RT }))
+            requestParams.RequestContext.Logger.Info(
+                "Filtering by authority specified in the authentication request parameters...");
+
+            ICollection<MsalAccessTokenCacheItem> authorityCacheMatches = tokenCacheItems.Where(
+                item => item.Authority.Equals(
+                    requestParams.Authority.CanonicalAuthority,
+                    StringComparison.OrdinalIgnoreCase)).ToList();
+
+            return authorityCacheMatches.Count > 0 ? tokenCacheItems : authorityCacheMatches;
```

Comment:
from @henrik-me in former [PR](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/708)
> i think this statement should be inverted... I wonder if the utests still passes... (alternatively you can say if Count > 0 then authorityMatches else tokenCacheItems …
...
seems like you changed both the validation and the return param... the U test should be catching this, is a utest missing?

I think using authorityMatches.Count > 0 makes sense. So if there are matches for the authority, we'll return both of them with the tokenCacheItems.
In the case I've been looking at, the authorities are not matching because the tenantId's are different. So the count is always 0, which means we'd return the tokenCacheItems, which might be 2 or 3, and we're back in the same loop as before, which is why I have it returning the authorityMatches and not the tokenCacheItems if the condition is not met. let's discuss further.